### PR TITLE
Sol/vault strategy connection

### PR DESCRIPTION
**Pull request by Sol3iman**
- Added more tests
- Refactor strategy contract
- Remove unused variables

### CODE REVIEW

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 1]
```diff
@@ -13,50 +13,42 @@ import "../libraries/CommonLibrary.sol";
...
     IPeriphery internal _periphery;
     IVAMM internal _vamm;
-    IVoltzVault.TickRange internal _currentPosition;
 
     // MUTABLE PARAMS
-
     uint256 internal _sigmaWad; // y (standard deviation parameter in wad 10^18)
-    int256 internal _max_possible_lower_bound_wad; // should be in fixed rate
-
+    int256 internal _maxPossibleLowerBoundWad; // should be in fixed rate
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: can change the comment on the right to smth more descriptive, e.g. "Maximum Possible Fixed Rate Lower bounds when initiating a rebalance", will make auditing the code more straightforward

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 2]
```diff
@@ -144,6 +111,9 @@ contract LPOptimiserStrategy is DefaultAccessControl {
     }
 
     /// @notice Get the nearest tick multiple given a tick and tick spacing
+    /// @param newTick The tick to be rounded to the closest multiple of tickSpacing
+    /// @param tickSpacing The tick spacing of the vamm being used for this strategy
+    /// @return int24 The nearest tick multiple for newTick
     function nearestTickMultiple(int24 newTick, int24 tickSpacing) public pure returns (int24) {
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol:121]: why does this function need to be public vs. internal?

Comment (arturbeg) [reply contracts/strategies/LPOptimiserStrategy.sol:121]: worth having unit tests for a couple of edge case scenarios for this function specifically if don't have already

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 3]
```diff
@@ -154,26 +124,27 @@ contract LPOptimiserStrategy is DefaultAccessControl {
 
     /// @notice Set new optimimal tick range based on current tick
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: small typo: "optimal"

Comment (arturbeg) [reply contracts/strategies/LPOptimiserStrategy.sol]: based on current twap tick given that we are using the offchain moving average of the fixed rate in the current iteration

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 4]
```diff
@@ -154,26 +124,27 @@ contract LPOptimiserStrategy is DefaultAccessControl {
...
     /// @param currentFixedRateWad currentFixedRate which is passed in from a 7-day rolling avg. historical fixed rate.
-    function rebalance(uint256 currentFixedRateWad) public returns (int24 newTickLowerMul, int24 newTickUpperMul) {
+    /// @return newTickLowerMul The new lower tick for the rebalanced position
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: i'd rename newTickLowerMul to newTickLower, same for upper tick

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 5]
```diff
@@ -154,26 +124,27 @@ contract LPOptimiserStrategy is DefaultAccessControl {
...
 
         // 1. Get the new tick lower
-        // write UTs to check for underflow
         int256 deltaWad = int256(currentFixedRateWad) - int256(_sigmaWad);
-        int256 newFixedLowerWad = 0;
-        if (deltaWad > 0) {
-            // delta is greater than 0 => choose delta
-            if (deltaWad < _max_possible_lower_bound_wad) {
+        int256 newFixedLowerWad;
+        if (deltaWad > 1e15) {
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: i'd put 1e15 into a constant at the top of the contract, the constant can be named "MINIMUM_FIXED_RATE"

Comment (arturbeg) [reply contracts/strategies/LPOptimiserStrategy.sol]: given 1e15 is right on the edge, is it worth making it e.g. 1e16 instead, i.e. 0.01 @0xZenus @Sol3iman ?

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 6]
```diff
@@ -154,26 +124,27 @@ contract LPOptimiserStrategy is DefaultAccessControl {
...
                 newFixedLowerWad = deltaWad;
             } else {
-                newFixedLowerWad = _max_possible_lower_bound_wad;
+                newFixedLowerWad = _maxPossibleLowerBoundWad;
             }
         } else {
-            // delta is less than or equal to 0 => choose 0
-            newFixedLowerWad = 0;
+            // delta is less than or equal to 1e15 (0.001) => choose 1e15 (0.001)
+            newFixedLowerWad = 1e15;
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: below on lines 153 and 158 where we do fixed rate to tick conversions isn't it cleaner to use the reverse logic of this https://github.com/Voltz-Protocol/voltz-core/blob/72a2b1f725ed53340046ab02de3eff4ba9cdd4bf/contracts/MarginEngine.sol#L155 @0xZenus which does not rely on prb math but rather relies on the low level math functions of v3

Comment (arturbeg) [reply contracts/strategies/LPOptimiserStrategy.sol]: tbh, might be worth packaging the fixed rate to tick conversion logic into a separate internal function that can then be separately tested in UTs with a few sample inputs

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 7]
```diff
@@ -190,20 +161,34 @@ contract LPOptimiserStrategy is DefaultAccessControl {
...
-        // _vault.rebalance(IVoltzVault.TickRange(_tickLower, _tickUpper));
+        _vault.rebalance(IVoltzVault.TickRange(newTickLowerMul, newTickUpperMul));
 
-        emit Rebalanced(newTickLowerMul, newTickUpperMul);
+        emit RebalancedTicks(newTickLowerMul, newTickUpperMul);
         return (newTickLowerMul, newTickUpperMul);
     }
+
+    /// @notice Callback function called after for ERC20RootVault::deposit
+    function depositCallback() external override {
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol:190]: interesting, why does the depositCallback need to be in the strategy, doesn't seem that intuitive to me architecturally @Sol3iman @0xZenus

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 2 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol:121]: I have written UTs for this function. The function should ideally be internal but if it is internal we can't test it. So that makes things a bit difficult from a UT perspective. We could make it internal and then have something in the readme to say "Change the nearestTickMultiple function to public for tests to work"

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 5 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: Agree. 1e16 would be better.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 6 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: That logic converts sqrtPrice to fixed rate. Here we need to convert the fixed rate to tick.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 7 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol:190]: The idea is as follows:
   - ERC20 Root Vault has sub-vaults and the underlying strategy (e.g. For Voltz: the subvaults are **erc20Vault** and **voltzVault** and the strategy is **LPOptimiserStrategy** -- the first subvault must be **erc20Vault**)
   - users deposit funds into ERC20 Root Vault and ERC20 Root Vault automatically transfers these funds to **erc20Vault** and then calls this callback. 
   - the callback is in strategy because that's the piece that dictates how funds are handled in the subvaults (i.e. For Voltz: the callback pulls all funds from **erc20Vault** to **voltzVault**)

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 6 above)
Comment (arturbeg) [reply contracts/strategies/LPOptimiserStrategy.sol]: yeah, that's true, can we not just do the reverse of that logic if that makes sense?
